ts/WoltLabSuite/Core/Api.ts

```diff
@@ -0,0 +1,294 @@
+  private _objectIDs: number[] = [];
+  private _payload: Payload = {};
+  private _showLoadingIndicator = true;
+  private _signal: AbortController | undefined = undefined;
+
+  private constructor(actionName: string, className: string) {
+    this.actionName = actionName;
+    this.className = className;
+  }
+
+  static dboAction(actionName: string, className: string): Api {
+    return new Api(actionName, className);
+  }
+
+  failure(failure: CallbackFailure): this {
+    this._failure = failure;
+
+    return this;
+  }
+
+  getAbortController(): AbortController {
+    if (this._signal === undefined) {
+      this._signal = new AbortController();
+    }
+
+    return this._signal;
+  }
+
+  objectIds(objectIds: number[]): this {
+    this._objectIDs = objectIds;
+
+    return this;
+  }
+
+  payload(payload: Payload): this {
+    this._payload = payload;
+
+    return this;
+  }
+
+  disableLoadingIndicator(): this {
+    this._showLoadingIndicator = false;
+
+    return this;
+  }
+
+  async dispatch(): Promise<unknown> {
+    const url = window.WSC_API_URL + "index.php?ajax-proxy/&t=" + Core.getXsrfToken();
+
+    const body: RequestBody = {
+      actionName: this.actionName,
+      className: this.className,
+    };
+    if (this._objectIDs) {
+      body.objectIDs = this._objectIDs;
+    }
+    if (this._payload) {
+      body.parameters = this._payload;
+    }
+
+    const init: RequestInit = {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
+        "X-Requested-With": "XMLHttpRequest",
+        "X-XSRF-TOKEN": Core.getXsrfToken(),
+      },
+      body: Core.serialize(body),
+      mode: "same-origin",
+      credentials: "same-origin",
+      cache: "no-store",
+      redirect: "error",
+    };
+
+    if (this._signal) {
+      init.signal = this._signal.signal;
+    }
+
+    // Use a local copy to isolte the behavior in case of changes before
+    // the request handling has completed.
+    const showLoadingIndicator = this._showLoadingIndicator;
+    if (showLoadingIndicator) {
+      AjaxStatus.show();
+    }
+
+    try {
+      const response = await fetch(url, init);
+
+      if (!response.ok) {
+        const result = this.handleError(ErrorCode.STATUS_NOT_OK, response);
+
+        return Promise.reject(result);
+      }
+
+      const contentType = response.headers.get("content-type");
+      if (!contentType || !contentType.includes("application/json")) {
+        const result = this.handleError(ErrorCode.EXPECTED_JSON, response);
+
+        return Promise.reject(result);
+      }
+
+      let json: ResponseData;
+      try {
+        json = await response.json();
+      } catch (e) {
+        const result = this.handleError(ErrorCode.INVALID_JSON, response);
+
+        return Promise.reject(result);
+      }
+
+      // This is an explicit wait for the event loop to execute the
+      // callee function before we execute additional tasks.
```

Comment:
I don't understand what this is attempting to say.

Reply:
You will often times see calls to `resolve()` without awaiting for them, but technically `void resolve()` would make it more clear that you don't care for its result.

Here we want to execute code after the action has run to mimic the existing behavior, so we need to chain it to `Promise.resolve()` which is a promise itself.